src/shared/sam/cli/samCliValidator.ts

```diff
@@ -0,0 +1,154 @@
+}
+
+export interface SamCliVersionValidatorResult {
+    version?: string,
+    validation: SamCliVersionValidation
+}
+
+export interface SamCliValidatorResult {
+    samCliFound: boolean
+    versionValidation?: SamCliVersionValidatorResult
+}
+
+export interface SamCliValidator {
+    detectValidSamCli(): Promise<SamCliValidatorResult>
+}
+
+export abstract class BaseSamCliValidator implements SamCliValidator {
+    private cachedSamInfoResponse?: SamCliInfoResponse
+    // The modification timestamp of SAM CLI is used as the "cache key"
+    private cachedSamInfoResponseSource?: Date
+
+    public constructor() {
+    }
+
+    public async detectValidSamCli(): Promise<SamCliValidatorResult> {
+        const result: SamCliValidatorResult = {
+            samCliFound: false
+        }
+
+        const samCliLocation = this.getSamCliLocation()
+        if (samCliLocation) {
+            result.samCliFound = true
+
+            result.versionValidation = await this.getVersionValidatorResult(samCliLocation)
+        }
+
+        return result
+    }
+
+    // This method is public for testing purposes
+    public async getVersionValidatorResult(samCliLocation: string): Promise<SamCliVersionValidatorResult> {
+        const cliStat: Pick<Stats, 'mtime'> = await this.getSamCliStat(samCliLocation)
+        if (!this.isSamCliVersionCached(cliStat.mtime)) {
+            this.cachedSamInfoResponse = await this.getInfo(samCliLocation)
+            this.cachedSamInfoResponseSource = cliStat.mtime
+        }
+
+        const version: string = this.cachedSamInfoResponse!.version
+
+        return {
+            version,
+            validation: BaseSamCliValidator.validateSamCliVersion(version),
+        }
+    }
+
+    protected abstract async getSamCliStat(samCliLocation: string): Promise<Pick<Stats, 'mtime'>>
+
+    protected abstract getSamCliLocation(): string | undefined
+
+    protected abstract async getInfo(samCliLocation: string): Promise<SamCliInfoResponse>
+
+    private isSamCliVersionCached(samCliLastModifiedOn: Date): boolean {
+        if (!this.cachedSamInfoResponse) { return false }
+        if (!this.cachedSamInfoResponseSource) { return false }
+
+        return this.cachedSamInfoResponseSource === samCliLastModifiedOn
+    }
+
+    public static validateSamCliVersion(version?: string): SamCliVersionValidation {
+        if (!version) {
+            return SamCliVersionValidation.VersionNotParseable
+        }
+
+        if (!semver.valid(version)) {
+            return SamCliVersionValidation.VersionNotParseable
+        }
+
+        if (semver.lt(version, MINIMUM_SAM_CLI_VERSION_INCLUSIVE)) {
+            return SamCliVersionValidation.VersionTooLow
+        }
+
+        if (semver.gte(version, MAXIMUM_SAM_CLI_VERSION_EXCLUSIVE)) {
+            return SamCliVersionValidation.VersionTooHigh
+        }
+
+        return SamCliVersionValidation.Valid
+    }
+}
+
+export class DefaultSamCliValidator extends BaseSamCliValidator {
```

Comment:
Do you envision any other extensions on `BaseSamCliValidator`? The functionality here seems fairly standard--is there a reason to have this pulled out into a different class?

Reply:
`BaseSamCliValidator` contains the logic so that it can be tested without using the actual sam cli.
See `src\test\shared\sam\cli\samCliValidator.test.ts` for corresponding tests.

Reply:
What do you think about extracting the communication with sam cli into a separate interface, rather abstract methods? So you would have:

```typescript
export interface SamCliContext {
    getInfo(): ...
}

export class DefaultSamCliValidator implements SamCliValidator {
    public constructor(private readonly context: SamCliContext) {
    }

    ...
    this.cachedSamInfoResponse = await this.context.getInfo(samCliLocation)
    ...
}
```

I would prefer this approach because:
1. This is the pattern that we use for this scenario in the rest of the codebase.
2. Composition is [more flexible and easier to reason about](https://en.wikipedia.org/wiki/Composition_over_inheritance) than inheritance in cases like this.
3. It clearly separates the two responsibilities (getting a response from SAM CLI, and validating that response).